### ISSUE TITLE
B300: fuse non-sparse Multi-Token Attention softmax backward

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/Liger-Kernel#7]
## Problem

The non-sparse Multi-Token Attention path already uses the shared Triton `_softmax_forward`, but discards its returned `BLOCK_SIZE`, `num_warps`, and launch metadata. Backward then reconstructs the same softmax derivative with eager PyTorch multiply, reduction, subtraction, and multiply operations:

```python
dot = (grad_probs * probs).sum(dim=-1, keepdim=True)
grad_scores_inf = probs * (grad_probs - dot)
```

`src/liger_kernel/ops/softmax.py` already provides the equivalent one-launch `_softmax_backward` used by `LigerSoftmaxFunction` and Fused Neighborhood Attention. Reusing it can remove intermediate tensors and GPU launches without adding or tuning a kernel.

This is a B300 performance hypothesis, not a claimed speedup. Shipped B200 BF16 data show a backward-specific Multi-Token Attention deficit: with B=2, C_in=C_out=4, K=3, Liger backward is 0.750784 ms versus torch 0.525184 ms at L=32 and 0.796320 versus 0.581824 ms at L=256, while forward is already faster at L=32. The shared softmax derivative is only one component, so B300 must be profile-gated before tracked changes.

Baseline: `upstream/main` at `124fb8a2a443c1118f7caeff0203aea78424eb2b`.

## Profile-first gate

Before any tracked source change, use one clean B300/SM103 Slurm allocation to compare the exact baseline non-sparse backward with a temporary variant that retains `_softmax_forward` metadata and invokes the existing `_softmax_backward`.

Use profiler/CUPTI evidence to verify that the eager softmax-gradient chain is actually present on this runtime and that the temporary variant replaces it with one Triton backward launch. Proceed only if the fixed reuse improves median backward or directly measured full latency by at least 10% on at least two representative non-sparse shapes, reduces softmax-gradient launches/intermediates as predicted, and has no greater than 5% regression in the gate matrix. Otherwise record a no-go without creating an empty draft PR or tuning another part of attention.

## Approved implementation scope

- In `src/liger_kernel/ops/multi_token_attention.py`, import and reuse the existing `_softmax_backward` only for `sparse=False`.
- Retain the `BLOCK_SIZE`, `num_warps`, and `multi_block_launch` values returned by the corresponding `_softmax_forward` on the autograd context and pass them to backward.
- Initialize/access that metadata only on the non-sparse branch so sparsemax behavior remains unchanged.
- Add focused correctness and regression coverage in `test/transformers/test_multi_token_attention.py`.

Expected tracked files:

- `src/liger_kernel/ops/multi_token_attention.py`
- `test/transformers/test_multi_token_attention.py`

Keep `src/liger_kernel/ops/softmax.py` and benchmark scripts unchanged. If upstream #1331 lands first, rebase mechanically and retain its independent device-context guards without absorbing other changes.

Explicitly out of scope:

- shared softmax kernel/configuration changes;
- sparsemax changes;
- causal mask forward/backward changes;
- convolution, grouping, or bias-gradient fusion/tuning;
- missing Multi-Token Attention features requested in upstream #857;
- Ascend/NPU work from upstream #1122, cuTile/CuTe backends, architecture dispatch, API changes, or benchmark-framework refactors.

## Correctness oracle

Compare functional and module paths with the existing `TorchMultiTokenAttention` reference. Use independent non-unit random upstream gradients so a constant-output sum cannot hide a softmax-Jacobian error. Compare forward output plus gradients for scores, weight, and bias.

Cover:

- FP16, BF16, and FP32 at repository tolerances;
- bias present and absent;
- groups 1 and a valid grouped/depthwise case;
- kernel sizes 1, 3, and 5;
- odd/ragged sequence lengths such as 5, 6, and 33 plus representative 64 and 128;
- non-contiguous public inputs as supported by `ensure_contiguous`.

Forward output must remain unchanged. Run the existing sparse=True cases as an unchanged-path control and require baseline parity for their output and gradients. Run the full Multi-Token Attention test file, the shared softmax test file, relevant functional tests, and repository checkstyle.

## B300 benchmark and acceptance

For final evidence, compare clean existing baseline and candidate commits in the same SM103 allocation using exact-source processes or clean worktrees. Reuse identical immutable inputs, weights, bias, and independent upstream gradients. Alternate provider order for at least seven rounds and report every round plus p20/median/p80.

Primary matrix:

- BF16 primary, with FP16 and FP32 controls;
- batch `1, 2`;
- input/output channels `4, 8`;
- sequence length `32, 64, 128, 256`;
- kernel size `3, 5`;
- groups `1, C_in` where shapes are valid;
- backward and directly measured full forward-plus-backward.

Include forward as a regression control and representative sparse=True cases as unchanged-path controls. Clear gradients and rebuild or safely retain graphs equivalently outside timing; do not derive full latency by adding independent forward/backward quantiles.

Profile representative BF16 L=64 and L=256 cases. Report softmax-backward kernel count, aggregate duration, intermediate allocations/bytes where the profiler exposes them, and contribution to full backward. The evidence should distinguish this bounded softmax change from the convolution cost that remains untouched.

Accept only if the final fixed reuse improves median backward or full latency by at least 10% on at least two representative non-sparse shapes, has no greater than 5% regression anywhere in the matrix or forward/sparse controls, and passes correctness. A narrow/noisy win is a no-go; do not broaden the scope to rescue it.

## Duplicate check

No open upstream issue, PR, remote branch, or fork artifact was found for reusing `_softmax_backward` in CUDA/Triton Multi-Token Attention. Upstream #857 requests missing MTA functionality, #1122 changes only the Ascend backend, and #1331 adds device contexts; none implements this optimization.

**Comment by heiheiha798** [comment]
## B300 result: no-go

The profile-first hypothesis reproduced, but the exact-commit final matrix failed the required no-regression gate. No draft PR was created.

### Candidate evaluated

- Baseline: `124fb8a2a443c1118f7caeff0203aea78424eb2b`
- Candidate: `52e603ee0c8fe4a835b823824e9316e4321010fd`
- Branch: `b300/mta-reuse-softmax-backward`
- Tracked diff: only `src/liger_kernel/ops/multi_token_attention.py` and `test/transformers/test_multi_token_attention.py`
- Candidate retained the three `_softmax_forward` launch values and reused the existing `_softmax_backward` only for `sparse=False`; shared softmax and sparse code were unchanged.

### Profile-first gate

Slurm job `4903` ran on an NVIDIA B300 SXM6 AC (SM103) from the clean baseline. The temporary reuse path cleared the initial gate:

- BF16, B=1, C=4, L=64, K=3, groups=1: backward `0.200595 -> 0.182090 ms` (`1.102x`), direct full `0.380253 -> 0.338426 ms` (`1.124x`).
- BF16, B=1, C=8, L=256, K=5, groups=1: backward `0.214227 -> 0.194458 ms` (`1.102x`).
- Isolated softmax-derivative allocations fell from `132096 -> 65536 bytes` at L=64 and `2101248 -> 1048576 bytes` at L=256.
- NCU showed the eager multiply/reduce/subtract/multiply sequence becoming one existing Triton softmax-backward launch.

This justified the bounded tracked implementation.

### Exact-commit final validation

Final Slurm job `4925` used clean detached worktrees for both commits and loaded each exact `multi_token_attention.py` into one Python/CUDA process under distinct module names. This removed the fixed-process/context bias observed in superseded jobs while preserving exact source attribution. Provider order alternated by shape and round.

Environment:

- Node: `dev-slurm`
- GPU: NVIDIA B300 SXM6 AC, UUID `GPU-3313f980-9de7-f4ee-9a38-4643f36424d3`, 148 SMs
- Driver: `580.126.09`; CUDA: `13.0`
- PyTorch: `2.13.0+cu130`; Triton: `3.7.1`; Nsight Compute: `2025.3.1`
- Command: `sbatch .cache/b300-mta-softmax-backward/exact_validation.sbatch`
- Matrix: 588 baseline/candidate comparisons, 7 alternating rounds, direct CUDA-event forward/backward/full timing, BF16/FP16/FP32, B={1,2}, C={4,8}, L={32,64,128,256}, K={3,5}, groups={1,C}, plus sparse controls.

Correctness/static results:

- `test/transformers/test_multi_token_attention.py`: `76 passed`
- `test/transformers/test_softmax.py`: `24 passed`
- Repository `make checkstyle`: passed on the login node with the repository virtual environment.

Mechanism evidence remained positive. Exact-source NCU measured:

- L=64: four eager kernels totaling `25.472 us` versus one `_softmax_single_block_backward_kernel` at `5.888 us`.
- L=256: four eager kernels totaling `26.176 us` versus one `_softmax_single_block_backward_kernel` at `6.656 us`.
- Full-MTA peak allocated memory was unchanged at `1706496 bytes` (L=64) and `77597184 bytes` (L=256), because unchanged convolution allocations dominate.

The performance acceptance nevertheless failed. To account for within-run clock drift, the gate was evaluated from each round's paired baseline/candidate ratio before taking the median. Eight non-sparse forward controls had median paired regressions greater than 5%, and each failed in four of seven rounds. Representative failures:

| dtype | B | C | L | K | groups | mode | paired median speedup |
|---|---:|---:|---:|---:|---:|---|---:|
| FP16 | 2 | 4 | 64 | 5 | 4 | forward | `0.916x` |
| BF16 | 1 | 4 | 128 | 5 | 4 | forward | `0.933x` |
| BF16 | 1 | 4 | 64 | 3 | 4 | forward | `0.934x` |
| BF16 | 2 | 8 | 64 | 3 | 8 | forward | `0.942x` |

Only four BF16/FP16 backward/full configurations retained a paired-median improvement of at least 10% in this most controlled run. Sparse controls were effectively unchanged (`0.987x` worst paired median), which supports attributing the short non-sparse forward cost to the additional autograd metadata bookkeeping rather than a general source-loader bias.

The candidate therefore violates the explicit requirement of no greater than 5% regression anywhere in the forward/sparse control matrix. Removing the metadata needed by the existing backward kernel, tuning another attention component, or relaxing the controls would exceed the approved problem. This cycle is recorded as `PR_NO_GO_RECORDED`; the branch and evidence are preserved for reference.

Raw evidence:

- `.cache/b300-mta-softmax-backward/exact-4925.out`
- `.cache/b300-mta-softmax-backward/exact-4925.err` (empty)
- `.cache/b300-mta-softmax-backward/exact-ncu-baseline-l64-4925.csv`
- `.cache/b300-mta-softmax-backward/exact-ncu-candidate-l64-4925.csv`
- `.cache/b300-mta-softmax-backward/exact-ncu-baseline-l256-4925.csv`
- `.cache/b300-mta-softmax-backward/exact-ncu-candidate-l256-4925.csv`

Jobs `4910`, `4911`, `4918`, and `4922` are superseded benchmark-method iterations and are not used for the terminal performance claim.
